### ISSUE TITLE
Colorful label fadeout issue.

### ISSUE THREAD

**Issue by fusijie** [cocos2d/cocos2d-x#11585]
`version:` v3.5Final
`issue:` Setting different colors to each letter of label, then run the fadeout action, the colors disappear.
`result:` 
![untitled](https://cloud.githubusercontent.com/assets/5085300/7313745/a135e8d0-ea8f-11e4-8366-bdac69240f39.gif)
`reproduce:` Add following code to HelloWorld.

``` cpp
    label->getLetter(0)->setColor(Color3B::RED);
    label->getLetter(1)->setColor(Color3B::GREEN);
    label->getLetter(2)->setColor(Color3B::BLUE);
    label->runAction(Sequence::create(DelayTime::create(2), FadeOut::create(2), nullptr));
```

**Comment by yangws** [comment]
you should call setCascadeOpacityEnable function first:

```
    label->setCascadeOpacityEnabled(true);
    label->getLetter(0)->setColor(Color3B::RED);
    label->getLetter(1)->setColor(Color3B::GREEN);
    label->getLetter(2)->setColor(Color3B::BLUE);
    label->runAction(Sequence::create(DelayTime::create(2), FadeOut::create(2), nullptr));
```

**Comment by fusijie** [comment]
OK, I see, thanks @yangws .

**Comment by WangPengJu** [comment]
cocos2d-x 3.2, this problem can't be solved. @yangws  What is your version? I use the same code.

```
    auto label = Label::createWithTTF("hello world", "fonts/Marker Felt.ttf", 40);
label->setPosition(480, 320);
label->setCascadeOpacityEnabled(true);
label->getLetter(0)->setColor(Color3B::RED);
label->getLetter(1)->setColor(Color3B::GREEN);
label->getLetter(2)->setColor(Color3B::BLUE);
label->runAction(Sequence::create(DelayTime::create(2), FadeOut::create(2), nullptr));
```

**Comment by WangPengJu** [comment]
It's work only in the new version? @fusijie  @zilongshanren  @pandamicro  Have you tested?

**Comment by WangPengJu** [comment]
英语翻译有点烂, 我还是说中文吧.见谅.  我现在用的版本的是3.2,这个问题我最早是在cocoachina提的,目前为止还是找不到解决的方法.看到上面的那个方法,我用相同的代码测试,也是不行.不知道 @yangws 是用什么版本.你们测试这个问题通过了吗?怎么直接把issue给关了.还是说在下个版本这样调用才生效?

**Comment by WangPengJu** [comment]
cocos2d-x  3.6 ,same problem.

**Comment by fusijie** [comment]
yes, It's not work in the v3 branch too. @yangws

**Comment by WangPengJu** [comment]
跟了一个下午,还好有源码可以修改.问题出在void Label::updateColor()里面,每个字的color保持不变,把alpha修改一下就可以了,不懂你们为什么覆盖掉颜色的值 -_-!!! 把如下代码修改一下就可以了:
 for (int index = 0; index < count; ++index)
        {
            quads[index].bl.colors = color4;
            quads[index].br.colors = color4;
            quads[index].tl.colors = color4;
            quads[index].tr.colors = color4;
            textureAtlas->updateQuad(&quads[index], index);
        }
 for (int index = 0; index < count; ++index)
        {
            quads[index].bl.colors.a = color4.a;
            quads[index].br.colors.a = color4.a;
            quads[index].tl.colors.a = color4.a;
            quads[index].tr.colors.a = color4.a;
            textureAtlas->updateQuad(&quads[index], index);
        }

**Comment by fusijie** [comment]
@WenhaiLin please review.

@pandamicro please reopen this issue.

**Comment by WangPengJu** [comment]
这样子改了,会导致正常label->setColor(Color3B)无效.. 郁闷中.

**Comment by a19352226** [comment]
已经解决了,加了个判断字符颜色是否被改变.谢谢各位大牛.
for (int index = 0; index < count; ++index)
        {
			std::set<int>::iterator itr = _changed.find(index);
			if (itr != _changed.end())
			{
				quads[index].bl.colors.a = color4.a;
				quads[index].br.colors.a = color4.a;
				quads[index].tl.colors.a = color4.a;
				quads[index].tr.colors.a = color4.a;
			}
			else
			{
				quads[index].bl.colors = color4;
				quads[index].br.colors = color4;
				quads[index].tl.colors = color4;
				quads[index].tr.colors = color4;
			}
            textureAtlas->updateQuad(&quads[index], index);
        }

**Comment by WangPengJu** [comment]
网络好卡, 一直发不上来. 我加了一个判断当前索引的字符是否被改变过,被改变过就alpha值改变,其他不变,否在直接覆盖. 项目可以用了, 在此谢过各位大牛的帮助.
![qq 20150715192008](https://cloud.githubusercontent.com/assets/9954852/8697380/28ad64d2-2b27-11e5-9f00-dbd089b32913.png)

**Comment by WenhaiLin** [comment]
Fixed at https://github.com/cocos2d/cocos2d-x/pull/12867

**Comment by WangPengJu** [comment]
3.2版本没有std::unordered_map<int, Sprite*> _letters; 唉,现在升级项目有点蛋疼.

**Comment by WenhaiLin** [comment]
@a19352226 额，版本跨度太大。或许你可以考虑比对同步ttf相关的以下几个文件：CCFontAtlas CCFontFreeType CCLabelTextFormatter CCLabel

**Comment by WenhaiLin** [comment]
@a19352226 3.2版本的话，你试试
```
label->setCascadeOpacityEnabled(true);
label->setCascadeColorEnabled(true);
```

**Comment by WangPengJu** [comment]
@WenhaiLin  那两个函数我试过了, 不管用. 依然谢谢你的解答. 我就用那个笨办法,实现需要的功能就好.  下个项目升级就不用这么麻烦了.

**Comment by wade0317** [comment]
> Hi, everybody.

I’m sorry to send Email to everybody directly!

I Known that cocos2d-x supports C++,Lua and JS, it do not support quick。But my project was developed by quick3.3.  because there is some bugs in quick 3.3 , such as emoji letter was not support very good. I found these bugs was resolved in coco2dx-3.7.  So i decided to trans our project from quick 3.3 to coco2dx-3.7。

I create a lua HelloWorld project by running "cocos new MyLuaGame -p com.yourcompany.mygame -l lua -d MyLuaGame" with the newest cocos2d-x  . I copy framework of quick 3.5 to my HelloWorld Project, and replace app directory , main.lua, config.lua with the sample code in mvc, when i run the project, it said : [LUA ERROR] [string "src/framework/cocos2dx/NodeEx.lua"]:278: attempt to call a nil value. I do not known why. Is that the newest cocos2d-x was not adapted to quick ? How can i fixed it ?  is the somebody known how to fixed it ?
